pylint/extensions/redefined_loop_name.py

```diff
@@ -0,0 +1,87 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/pylint/blob/main/CONTRIBUTORS.txt
+
+"""Optional checker to warn when loop variables are overwritten in the loop's body."""
+
+from __future__ import annotations
+
+from astroid import nodes
+
+from pylint import checkers
+from pylint.checkers import utils
+from pylint.interfaces import HIGH
+from pylint.lint import PyLinter
+
+
+class RedefinedLoopName(checkers.BaseChecker):
+
+    name = "redefined-loop-name"
+
+    msgs = {
+        "W2901": (
+            "Redefining %r from loop (line %s)",
+            "redefined-loop-name",
+            "Used when a loop variable is overwritten in the loop body.",
+        ),
+    }
+
+    def __init__(self, linter: PyLinter) -> None:
+        super().__init__(linter)
+        self._loop_variables: list[tuple[nodes.For, list[str | None]]] = []
+
+    @utils.only_required_for_messages("redefined-loop-name")
+    def visit_assignname(self, node: nodes.AssignName) -> None:
+        assign_type = node.assign_type()
+        if not isinstance(assign_type, (nodes.Assign, nodes.AugAssign)):
+            return
+        node_scope = node.scope()
+        for outer_for, outer_variables in self._loop_variables:
+            if node_scope is not outer_for.scope():
```

Comment:
Should we add the `scope` to `_loop_variables`? That could be expensive to call every time?

Reply:
Hmm. Parent lookups are O(1), right? That doesn't seem too bad. As opposed to iterating over a large body of statements or all children.

Reply:
Yeah it is definitely not the most expensive call, but since we're doing it on every pass I thought it might make sense. I'll leave this to you though!

Reply:
Looking at this more closely I think you're right -- thanks for the idea 👍🏻